tests/src/test/java/io/streamnative/pulsar/handlers/kop/TransactionTest.java

```diff
@@ -379,11 +384,134 @@ private KafkaProducer<Integer, String> buildIdempotenceProducer() {
+        producer1.close();
+        producer2.close();
+    }
+
+    @Test
+    public void testProducerFencedWhileSendOffsets() throws Exception {
+        final KafkaProducer<Integer, String> producer1 = buildTransactionProducer("prod-1");
+        producer1.initTransactions();
+        producer1.beginTransaction();
+        producer1.send(new ProducerRecord<>("test", "test"))
+                .get();
+
+        final KafkaProducer<Integer, String> producer2 = buildTransactionProducer("prod-1");
+        producer2.initTransactions();
+        producer2.beginTransaction();
+        producer2.send(new ProducerRecord<>("test", "test")).get();
+
+
+        // producer1 cannot offsets
+        expectThrows(ProducerFencedException.class, () -> {
+            producer1.sendOffsetsToTransaction(ImmutableMap.of(new TopicPartition("test", 0),
+                            new OffsetAndMetadata(0L)),
+                    "testGroup");
+        });
+
+        // and it cannot commit
+        expectThrows(ProducerFencedException.class, () -> {
+            producer1.commitTransaction();
+        });
+
+        producer1.close();
+        producer2.close();
+    }
+
+    @Test
+    public void testProducerFencedWhileAbortAndBegin() throws Exception {
+        final KafkaProducer<Integer, String> producer1 = buildTransactionProducer("prod-1");
+        producer1.initTransactions();
+        producer1.beginTransaction();
+        producer1.send(new ProducerRecord<>("test", "test"))
+                .get();
+
+        final KafkaProducer<Integer, String> producer2 = buildTransactionProducer("prod-1");
+        producer2.initTransactions();
+        producer2.beginTransaction();
+        producer2.send(new ProducerRecord<>("test", "test")).get();
+
+        // producer1 cannot abort
+        expectThrows(ProducerFencedException.class, () -> {
+            producer1.abortTransaction();
+        });
+
+        // producer1 cannot start a new transaction
+        expectThrows(ProducerFencedException.class, () -> {
+            producer1.beginTransaction();
+        });
+        producer1.close();
+        producer2.close();
+    }
+
+    @Test
+    public void testNotFencedWithBeginTransaction() throws Exception {
+        final KafkaProducer<Integer, String> producer1 = buildTransactionProducer("prod-1");
+        producer1.initTransactions();
+
+        final KafkaProducer<Integer, String> producer2 = buildTransactionProducer("prod-1");
+        producer2.initTransactions();
+        producer2.beginTransaction();
+        producer2.send(new ProducerRecord<>("test", "test")).get();
+
+        // beginTransaction doesn't do anything
+        producer1.beginTransaction();
+
+        producer1.close();
+        producer2.close();
+    }
+
     /**
      * Get the Kafka server address.
      */
     private String getKafkaServerAdder() {
-        return "localhost:" + getKafkaBrokerPort();
+        return "localhost:" + getClientPort();
```

Comment:
this change is not strictly needed, but it helps me as all the Proxy tests use getClientPort()
I would send a separate PR to replace all the getKafkaBrokerPort(); with getClientPort()